src/explore-education-statistics-frontend/src/pages/api/permalink.ts

```diff
@@ -0,0 +1,72 @@
+import mapFullTable from '@common/modules/table-tool/utils/mapFullTable';
+import mapTableHeadersConfig from '@common/modules/table-tool/utils/mapTableHeadersConfig';
+import mapTableToJson from '@common/modules/table-tool/utils/mapTableToJson';
+import generateTableTitle from '@common/modules/table-tool/utils/generateTableTitle';
+import type { NextApiRequest, NextApiResponse } from 'next';
+
+interface ResponseData {
+  table?: string;
+  title?: string;
+  error?: unknown;
+}
+
+/**
+ * Endpoint to generate table json and title for permalinks.
+ *
+ * Expects req.body to be:
+ * {
+ *  fullTable: TableDataResponse,
+ *  configuration: {
+ *   tableHeaders: UnmappedTableHeadersConfig
+ *  }
+ * }
+ */
+export default async function handlePermalink(
+  req: NextApiRequest,
+  res: NextApiResponse<ResponseData>,
+) {
+  if (
+    process.env.NODE_ENV === 'production' &&
+    req.query.PERMALINK_API_SECRET !== process.env.PERMALINK_API_SECRET
+  ) {
+    res.status(401).send({ error: 'Unauthorised' });
+    return;
+  }
+
+  if (req.method === 'POST') {
```

Comment:
Perhaps we could refactor the handling of methods into a higher-order function. Be interesting to hear other poeple's thoughts on this. I've found it just reduces noise inside the function body due to not having `if` statements to handle whether or not the right method is being supplied. 

This function would look something like: 

```typescript
import type { NextApiHandler, NextApiRequest, NextApiResponse } from 'next';

type Method = 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE' | 'OPTIONS' | 'HEAD';

export default function withMethods(methods: Method[], handler: NextApiHandler) {
  // eslint-disable-next-line func-names
  return async function (req: NextApiRequest, res: NextApiResponse) {
    if (!methods.includes(req.method as Method)) {
      return res.status(405).end();
    }

    return handler(req, res);
  };
}

```

and would be used like: 

```typescript 
const handler = async (req: NextApiRequest, res: NextApiResponse) => {
  // ...
};
export default withMethods(['GET'], handler);
```

Reply:
As this is a temporary endpoint that can be removed once the permalinks migration is completed then it seems a bit unnecessary to add middleware for it which will be unused once the endpoint is removed. Unless we expect that we'll be adding other endpoints in future?

Reply:
From what we've discussed previously, I think this endpoint will still be used even after the migration is complete. The flow would look something like this:


```mermaid
sequenceDiagram
    Browser->>Content API: POST /api/permalink
    Content API->>Next.js: POST /api/permalink
    Next.js-->>Content API: Rendered permalink table
    Content API->>Blob Storage: Uploads permalink table
    Content API-->>Browser: Rendered permalink table
```

I think [EES-3792](https://dfedigital.atlassian.net/browse/EES-3792) documented an approach you might be thinking of. We had 3 Amigo's a couple of weeks ago and we changed various things in it to reflect the above.

It'd also be good to have some of this infrastructure in place for the next time we need Next.js endpoints as well.

Reply:
Ah ok, I didn't realise the plan had changed from when Ben and I originally wrote the tickets (ages ago now!).

Reply:
Middleware added now.